29 gamecontroller

stephensmeal opened the pull request:
PR to merge game controller implementation into integ

stephensmeal committed:
Implementation and Unit Tests for GameController Class

stephensmeal committed:
Updated Still In Progress GameController Implementation

gamecontroller.py - Updated implementation
gamestate.py - Bug fix in start() to set correct first player
gamestatus.py - Added and remove GameStatus(es) for functionality

stephensmeal committed:
Bug Fix In GameState And Fix Unit Tests

Previous change to gamestate.py start() method needed to be
fixed to call self.next_turn(). GameState unit test
(test_gamestate.py) needed to be changed to reflect updated
GameStatus(es).

stephensmeal committed:
Incremental GameController and Test Implementation

patmay commented on src/gamecontroller.py:
```diff
@@ -9,20 +19,137 @@ class GameController(object):
+                    if selected_player.get_uuid() is None:
+                        selected_player.set_uuid(msg_uuid)
+                        update_message = Message(sni.get_uuid(), MessageType.SEND_PLAYERS, pl.serialize())
+
```
Bump update_message outside of the if statement, and use SNI to send it

patmay commented on src/gamecontroller.py:
```diff
@@ -9,20 +19,137 @@ class GameController(object):
+                        selected_player.set_uuid(msg_uuid)
+                        update_message = Message(sni.get_uuid(), MessageType.SEND_PLAYERS, pl.serialize())
+
+                if msg_type == MessageType.START_GAME:
```
Check to see if you have enough players

patmay commented on src/gamecontroller.py:
```diff
@@ -9,20 +19,137 @@ class GameController(object):
+
+                elif msg_type == MessageType.SUGGESTION_MAKE:
+                    suggestion = Suggestion.deserialize(msg_payload)
+                    self._suggest_engine.make_suggestion(suggestion)
```
check function output of make suggestion.

patmay commented on src/gamecontroller.py:
```diff
@@ -9,20 +19,137 @@ class GameController(object):
+    def do_accusation(msg_uuid: str, msg_type: MessageType, msg_payload: str):
+        accusation = Suggestion.deserialize(msg_payload)
+        pl = PlayerList()
+        accuser = pl[self._current_game.get_current_player()]
```
have to do 
list = pl.get_players()
accuser = list[self._current_game.get_current_player()]

stephensmeal committed:
Implementation of GameController and Bug Fixes

gamecontroller.py - Addresses Comments and Added Helper Functions
suggest_engine.py - Added is_valid_suggestion() function to ensure
    suggesting player is in suggested room
test_gamecontroller.py - Removed from branch to increase cycle time

stephensmeal committed:
Fixed Pep8 Issues in GameController

gamecontroller.py - Fixed Pep8 Issues

patmay commented on src/gamecontroller.py:
```diff
@@ -9,20 +22,176 @@ class GameController(object):
+        """
+        if msg_type == MessageType.LEAVE_GAME:
+            current_player_index = self._current_game.get_current_player()
+            current_player_uuid = pl[current_player_index].get_uuid()
```
I don't think you can do this because pl is not an array. You should be able to do `pl.get_player_by_index(index)` now.

stephensmeal replied:
This needs to be refactored completely. UUID is in the Message, that can be used to pull player, etc. I'll work on it.

patmay commented on src/gamecontroller.py:
```diff
@@ -9,20 +22,176 @@ class GameController(object):
+                    if total_players > 2:
+                        self._current_game.start()
+                        self._current_game.set_state(GameStatus.START_TURN)
+
```
We need a to send people's dealt cards to them. we probably need to do Hand.serialize() and iterate though the playerlist. this might be a job for GameState. would like to hear your thoughts on where it goes.

stephensmeal replied:
Originally we talked about the player's hand's cards being automatically checkout on the notecard. Sending a message is good as well, but since we're only allowing four lines in the message center it could move out before the play marks their own card and there is no way for the player to recall their cards again.

patmay replied:
Whether we put a message in the message center, or we check the notecard for them, we still need to send a message with their hand

stephensmeal replied:
Should this be at the end of GameState start()?

patmay commented on src/gamecontroller.py:
```diff
@@ -9,20 +22,176 @@ class GameController(object):
+                    move = Move.deserialize(msg_payload)
+                    if self._move_engine.is_valid_move(move):
+                        self._move_engine.do_move(move)
+                        self._current_game.set_state(GameStatus.POST_MOVE)
```
need to change Player's was_transferred to False

patmay commented on src/gamecontroller.py:
```diff
@@ -9,20 +22,176 @@ class GameController(object):
+            suggesting_player.set_was_transferred(False)
+            update_board_message = Message(
+                sni.get_uuid(), MessageType.NOTIFY, Board.serialize())
+            sni.send_all(update_board_message)
```
This all looks really good.

patmay commented on src/gamestate.py:
```diff
@@ -104,7 +104,8 @@ def start(self):
 
         # Reset self._current_player
-        self._current_player = 0
+        self._current_player = 5
```
would prefer this to be self._current_player = len(playerlist)

stephensmeal replied:
That's fine but I think it should be len(playerlist) - 1. len(playerlist) should return 6 and would rather not use an out of bounds index.

patmay commented on src/gamecontroller.py:
```diff
@@ -9,20 +22,176 @@ class GameController(object):
+        accuser = players[self._current_game.get_current_player()]
+        if self._suggest_engine.make_accusation(accusation, accuser):
+            self._current_game.set_state(GameStatus.LOBBY)
+        else:
```
need to change player's status to lost or whatever it is

stephensmeal replied:
Good catch.

patmay commented on src/gamecontroller.py:
```diff
@@ -9,20 +22,176 @@ class GameController(object):
+        self._current_game.set_state(GameStatus.START_TURN)
+
+    def set_next_player():
+        next_player = self._current_game.next_turn()
```
gamestate.next_turn() updates the current player. set_next_player should probably be removed and replaced with gamestate.next_turn()

stephensmeal replied:
Word.

yo09975 commented:
Bumping back to in progress until comments are addressed.

stephensmeal committed:
GameController and GameState Updates

gamecontroller.py - Addresses PR comments and made sure that
moving players' was_transferred flag is set to False
gamestate.py - Addresses PR comments and fixed bug in next_turn
which would cause error if no it was called with no active
players remaining in PlayerList.

patmay commented on src/gamecontroller.py:
```diff
@@ -9,20 +22,179 @@ class GameController(object):
+                if msg_type == MessageType.START_GAME:
+                    total_players = 0
+                    for p in pl:
+                        if p.get_status() == PlayerStatus.COMP:
```
Shouldn't this be PlayerStatus.ACTIVE? Should we set the PlayerStatus in SELECT_PIECE?

should be `for p in pl.get_players()`

stephensmeal replied:
We really should have called PlayerList something other than list because it just keeps getting fucked up in my brain.

patmay commented on src/gamecontroller.py:
```diff
@@ -9,20 +22,179 @@ class GameController(object):
+                        self._move_engine.do_move(move)
+                        moving_player = self.get_player_from_uuid(msg_uuid)
+                        moving_player.set_was_transferred(False)
+                        self._current_game.set_state(GameStatus.POST_MOVE)
```
Need to serialize board and send to all

patmay commented on src/gamecontroller.py:
```diff
@@ -9,20 +22,179 @@ class GameController(object):
+            suggested_player.set_was_transferred(True)
+            suggesting_player.set_was_transferred(False)
+            update_board_message = Message(
+                sni.get_uuid(), MessageType.NOTIFY, Board.serialize())
```
Board is not a singleton. You can access it (illegally) with `self._move_engine._board.serialize()`

stephensmeal replied:
No getter in MovementEngine?

patmay commented on src/gamecontroller.py:
```diff
@@ -83,14 +83,15 @@ def read_message(self, message: Message):
-                        if p.get_status() == PlayerStatus.COMP:
+                    for p in pl.get_players():
+                        if p.get_status() == PlayerStatus.ACTIVE:
                             total_players += 1
```
i don't think you can do += in python

yo09975 commented:
Needs to be rebased

stephensmeal committed:
Rebasing on integ
